File: algorithms/dijkstra.py
from algorithms import algorithm
import heapq

class Dijkstra(algorithm.Algorithm):
    def __init__(self, graph, start_node, end_node):
        super().__init__(graph, start_node, end_node)
        self.distances = {i: float('inf') for i in range(len(graph))}
        self.distances[start_node] = 0
        self.came_from = {start_node: None}
        self.pq = [(0, start_node)]
        self.closed_set = set()

    def solve(self):
        self.frontier.add(self.start_node)
        yield self.frontier, self.closed_set, self.start_node, [self.start_node]

        while self.pq:
            current_distance, current_node = heapq.heappop(self.pq)
            self.frontier.discard(current_node)

            if current_node in self.closed_set:
                continue
                
            self.closed_set.add(current_node)

            if current_node == self.end_node:
                yield self.frontier, self.closed_set, current_node, self.reconstruct_path()
                return

            for neighbor in range(len(self.graph[current_node])):
                if self.graph[current_node][neighbor] > 0: # Connected
                    weight = self.graph[current_node][neighbor]
                    distance = current_distance + weight

                    if distance < self.distances[neighbor]:
                        self.distances[neighbor] = distance
                        self.came_from[neighbor] = current_node
                        heapq.heappush(self.pq, (distance, neighbor))
                        self.frontier.add(neighbor)
            
            yield self.frontier, self.closed_set, current_node, self.reconstruct_path(current_node)

    def reconstruct_path(self, target_node=None):
        if target_node is None:
            target_node = self.end_node
        path = []
        current_node = target_node
        while current_node is not None:
            path.append(current_node)
            current_node = self.came_from.get(current_node)
        path.reverse()
        return path

File: src/app/core/algorithms/dijkstra.ts
import { Algorithm, NodeState } from './algorithm';

export class Dijkstra extends Algorithm {
  private closedSet: Set<number> = new Set();
  private frontierSet: Set<number> = new Set();
  private pq: { distance: number, node: number }[] = [];
  private cameFrom: Map<number, number> = new Map();
  private distances: Map<number, number> = new Map();

  constructor(graph: number[][], startNode: number, endNode: number) {
    super(graph, startNode, endNode);

    for (let i = 0; i < graph.length; i++) {
      this.distances.set(i, Infinity);
    }

    this.distances.set(startNode, 0);
    this.pq.push({ distance: 0, node: startNode });
  }

  *solve(): Generator<NodeState, void, unknown> {
    this.frontierSet.add(this.startNode);
    yield {
      frontier: new Set(this.frontierSet),
      visited: new Set(this.closedSet),
      currentNode: this.startNode,
      path: [this.startNode]
    };

    while (this.pq.length > 0) {
      this.pq.sort((a, b) => a.distance - b.distance);
      const current = this.pq.shift()!;
      const currentDistance = current.distance;
      const currentNode = current.node;

      this.frontierSet.delete(currentNode);

      if (this.closedSet.has(currentNode)) {
        continue;
      }

      this.closedSet.add(currentNode);

      if (currentNode === this.endNode) {
        yield {
          frontier: new Set(this.frontierSet),
          visited: new Set(this.closedSet),
          currentNode: currentNode,
          path: this.reconstructPath()
        };
        return;
      }

      for (let neighbor = 0; neighbor < this.graph[currentNode].length; neighbor++) {
        const weight = this.graph[currentNode][neighbor];
        if (weight > 0) {
          const distance = currentDistance + weight;

          if (distance < this.distances.get(neighbor)!) {
            this.distances.set(neighbor, distance);
            this.cameFrom.set(neighbor, currentNode);
            this.pq.push({ distance, node: neighbor });
            this.frontierSet.add(neighbor);
          }
        }
      }

      yield {
        frontier: new Set(this.frontierSet),
        visited: new Set(this.closedSet),
        currentNode: currentNode,
        path: this.reconstructPath(currentNode)
      };
    }
  }

  private reconstructPath(targetNode: number | null = null): number[] {
    let current = targetNode !== null ? targetNode : this.endNode;
    const path: number[] = [];
    while (this.cameFrom.has(current)) {
      path.push(current);
      current = this.cameFrom.get(current)!;
    }
    path.push(this.startNode);
    return path.reverse();
  }
}
